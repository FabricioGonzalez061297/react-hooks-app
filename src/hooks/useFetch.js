import { useState, useEffect, useRef } from "react"

export const UseFetch = (url) => {

    const isMounted = useRef(true)

    const [state, setState] = useState({
        data: null,
        isLoading: true,
        hasError: null,
        msgError: null
    });


    useEffect(()=>{

        return ()=>{
            isMounted.current=false;
        }
    },[])

    useEffect(() => {

        setState({ isLoading: true, data: null, msgError: null, hasError: false });

        fetch(url).then(resp => resp.json())
            .then(result => {

                if(isMounted.current){
                    setState({ data: result, isLoading: false, hasError: false, msgError: null });
                }
            })
            .catch(err => {
                if(isMounted.current){
                    setState({ data: null, isLoading: false, hasError: true, msgError:'ocurrio un error' });
                }

            })

    }, [url]);



    return state;

}