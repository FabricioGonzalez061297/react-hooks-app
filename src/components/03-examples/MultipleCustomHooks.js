import React from 'react';

import './multipleCustomHooks.css';
import { UseFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

export const MultipleCustomHooks = () => {

    const {state:contador,increment} = useCounter(8);
    const { data, isLoading } = UseFetch(`https://www.breakingbadapi.com/api/quotes/${contador}`);

    const {author,quote} = !!data && data[0];


    return (
        <>
            <h1>Breakingbad quotes</h1>
            <hr></hr>

            {
                isLoading
                    ?
                    (
                        <div className="alert alert-info text-center">
                            loading....
                        </div>
                    )
                    :
                    (
                        <blockquote className="blockquote text-right" >
                            <p className="mb-0"> {quote} </p>
                            <footer className="blockquote-footer"> {author} </footer>
                        </blockquote>

                    )
            }

            <br></br>

            <button
                disabled={isLoading}
                onClick={()=>increment(1)}
                className="btn btn-primary"
                type="button"
            >Siguiente.</button>

        </>
    );
}