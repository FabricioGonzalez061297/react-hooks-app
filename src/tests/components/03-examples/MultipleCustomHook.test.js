import React from 'react';
import { shallow } from 'enzyme';
import { MultipleCustomHooks } from '../../../components/03-examples/MultipleCustomHooks';
import { UseFetch } from '../../../hooks/useFetch';
import { useCounter } from '../../../hooks/useCounter';


//no interesa la implementación sino que valores retorna 

//en ves de utilizar el useFetch en este archivo, va a utilizar una implementación propia
jest.mock('../../../hooks/useFetch');
jest.mock('../../../hooks/useCounter');


describe('Pruebas en <MultipleCustomHooks/>', () => {

    beforeEach(()=>{
        
        useCounter.mockReturnValue({
            state:10,
            increment:()=>{},
            decrement:()=>{},
            reset:()=>{}
        });
    })


    test('debe de mostrarse correctamente', async () => {

        UseFetch.mockReturnValue({
            data: null,
            isLoading: true,
            hasError: null,
            msgError: null
        });


        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper).toMatchSnapshot();
    });

    test('debe de mostrar la informacion', async () => {

        UseFetch.mockReturnValue({
            data: [{
                author:'Fabricio',
                quote:'hola mundo'
            }],
            isLoading: false,
            hasError: false,
            msgError: null
        });

        const wrapper = shallow(<MultipleCustomHooks />);

        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe('hola mundo');

    });

});