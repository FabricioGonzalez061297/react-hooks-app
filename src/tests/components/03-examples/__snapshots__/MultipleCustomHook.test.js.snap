// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <MultipleCustomHooks/> debe de mostrarse correctamente 1`] = `
<Fragment>
  <h1>
    Breakingbad quotes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    loading....
  </div>
  <br />
  <button
    className="btn btn-primary"
    disabled={true}
    onClick={[Function]}
    type="button"
  >
    Siguiente.
  </button>
</Fragment>
`;
